src/mbgl/text/bidi.cpp

```diff
@@ -2,55 +2,123 @@
 
 #include <mbgl/text/bidi.hpp>
 #include <unicode/ubidi.h>
-#include <unicode/ubiditransform.h>
 #include <unicode/ushape.h>
 
 namespace mbgl {
 
-BiDi::BiDi() {
+// Takes UTF16 input in logical order and applies Arabic shaping to the input while maintaining
+// logical order
+//  Output won't be intelligible until the bidirectional algorithm is applied
+std::u16string applyArabicShaping(const std::u16string& input) {
     UErrorCode errorCode = U_ZERO_ERROR;
-    transform = ubiditransform_open(&errorCode); // Only error is failure to allocate memory, in
-                                                 // that case ubidi_transform would fall back to
-                                                 // creating transform object on the fly
+
+    int32_t outputLength =
+        u_shapeArabic(input.c_str(), static_cast<int32_t>(input.size()), NULL, 0,
+                      (U_SHAPE_LETTERS_SHAPE & U_SHAPE_LETTERS_MASK) |
+                          (U_SHAPE_TEXT_DIRECTION_LOGICAL & U_SHAPE_TEXT_DIRECTION_MASK),
+                      &errorCode);
+
+    // Pre-flighting will always set U_BUFFER_OVERFLOW_ERROR
+    errorCode = U_ZERO_ERROR;
+
+    std::unique_ptr<UChar[]> outputText = std::make_unique<UChar[]>(outputLength);
+    u_shapeArabic(input.c_str(), static_cast<int32_t>(input.size()), outputText.get(), outputLength,
+                  (U_SHAPE_LETTERS_SHAPE & U_SHAPE_LETTERS_MASK) |
+                      (U_SHAPE_TEXT_DIRECTION_LOGICAL & U_SHAPE_TEXT_DIRECTION_MASK),
+                  &errorCode);
+
+    // If the algorithm fails for any reason, fall back to non-transformed text
+    if (U_FAILURE(errorCode))
+        return input;
+
+    return std::u16string(outputText.get(), outputLength);
+}
+
+ProcessedBiDiText::ProcessedBiDiText(BiDi& p_bidi) : bidi(p_bidi) {
+}
+
+void ProcessedBiDiText::mergeParagraphLineBreaks(std::set<int32_t>& lineBreakPoints) {
+    for (int32_t i = 0; i < ubidi_countParagraphs(bidi.bidiText); i++) {
```

Comment:
Can/should `ubidi_countParagraphs(bidi.bidiText)` be extracted from the loop condition, so it's executed only once?